boa_engine/src/vm/code_block.rs

```diff
@@ -625,41 +604,116 @@ pub(crate) fn create_function_object(
         )
     };
 
-    let constructor =
-        JsObject::from_proto_and_data(function_prototype, ObjectData::function(function));
+    let data = ObjectData::function(function);
 
-    let constructor_property = PropertyDescriptor::builder()
-        .value(constructor.clone())
-        .writable(true)
-        .enumerable(false)
-        .configurable(true)
-        .build();
+    let templates = context.intrinsics().templates();
 
-    constructor
-        .define_property_or_throw(utf16!("length"), length_property, context)
-        .expect("failed to define the length property of the function");
-    constructor
-        .define_property_or_throw(utf16!("name"), name_property, context)
-        .expect("failed to define the name property of the function");
+    let (mut template, storage, constructor_prototype) = if r#async || arrow || method {
+        (
+            templates.function_without_proto().clone(),
+            vec![length, name],
+            None,
+        )
+    } else {
+        let constructor_prototype = templates
+            .function_prototype()
+            .create(ObjectData::ordinary(), vec![JsValue::undefined()]);
 
-    if !r#async && !arrow && !method {
-        let prototype = JsObject::with_object_proto(context.intrinsics());
-        prototype
-            .define_property_or_throw(CONSTRUCTOR, constructor_property, context)
-            .expect("failed to define the constructor property of the function");
-
-        let prototype_property = PropertyDescriptor::builder()
-            .value(prototype)
-            .writable(true)
-            .enumerable(false)
-            .configurable(false)
-            .build();
-        constructor
-            .define_property_or_throw(PROTOTYPE, prototype_property, context)
-            .expect("failed to define the prototype property of the function");
+        let template = templates.function_with_prototype_without_proto();
+
+        (
+            template.clone(),
+            vec![length, name, constructor_prototype.clone().into()],
+            Some(constructor_prototype),
+        )
+    };
+
+    template.set_prototype(prototype);
+
+    let contructor = template.create(data, storage);
+
+    if let Some(constructor_prototype) = &constructor_prototype {
+        constructor_prototype.borrow_mut().properties_mut().storage[0] = contructor.clone().into();
     }
+    contructor
+}
 
-    constructor
+pub(crate) fn create_function_object_fast(
```

Comment:
Could we add some documentation to this function, to understand the difference with `create_function_object()`? Also, in `create_function_object()` we should explain when is this one preferred.